### PR TITLE
feat(esm): upgrading oas tooling deps for esm support

### PR DESCRIPTION
**Pull request by erunion**
## 🧰 Changes

This updates the following dependencies for their newly packaged ESM + CJS export support:

* `@readme/data-urls`
* `@readme/oas-extensions`
* `oas`
* `remove-undefined-objects`

### CODE REVIEW

**Comment by erunion** [comment]
@prisis I'm admittedly out of my element on this one, but any idea why the `oas` updates you did aren't working here?

Screen Shot 2023-09-01 at 4 51 21 PM

If I change these `oas/dist/rmoas.types` imports to be `oas/rmoas.types` the tests pass but then TypeScript can no longer find the type definitions for it:

Screen Shot 2023-09-01 at 4 52 21 PM

```
> npx tsc
src/index.ts:16:8 - error TS2307: Cannot find module 'oas/rmoas.types' or its corresponding type declarations.

16 } from 'oas/rmoas.types';
          ~~~~~~~~~~~~~~~~~

src/index.ts:22:23 - error TS2307: Cannot find module 'oas/rmoas.types' or its corresponding type declarations.

22 import { isRef } from 'oas/rmoas.types';
                         ~~~~~~~~~~~~~~~~~

src/lib/configure-security.ts:1:56 - error TS2307: Cannot find module 'oas/rmoas.types' or its corresponding type declarations.

1 import type { OASDocument, SecuritySchemeObject } from 'oas/rmoas.types';
                                                         ~~~~~~~~~~~~~~~~~

src/lib/configure-security.ts:3:23 - error TS2307: Cannot find module 'oas/rmoas.types' or its corresponding type declarations.

3 import { isRef } from 'oas/rmoas.types';
                        ~~~~~~~~~~~~~~~~~

src/lib/style-formatting/index.ts:2:52 - error TS2307: Cannot find module 'oas/rmoas.types' or its corresponding type declarations.

2 import type { ParameterObject, SchemaObject } from 'oas/rmoas.types';
                                                     ~~~~~~~~~~~~~~~~~

src/lib/utils.ts:1:47 - error TS2307: Cannot find module 'oas/rmoas.types' or its corresponding type declarations.

1 import type { JSONSchema, SchemaObject } from 'oas/rmoas.types';
                                                ~~~~~~~~~~~~~~~~~


Found 6 errors in 4 files.

Errors  Files
     2  src/index.ts:16
     2  src/lib/configure-security.ts:1
     1  src/lib/style-formatting/index.ts:2
     1  src/lib/utils.ts:1
```

Any ideas what I'm doing wrong? Appreciate your help.

**Comment by erunion** [comment]
Managed to rubberduck it with this tsconfig change:

```diff
diff --git a/tsconfig.json b/tsconfig.json
index f1c5db0..c086df6 100644
--- a/tsconfig.json
+++ b/tsconfig.json
@@ -6,7 +6,8 @@
     "downlevelIteration": true,
     "esModuleInterop": true,
     "lib": ["DOM", "ES2020"],
-    "moduleResolution": "node",
+    "module": "Node16",
+    "moduleResolution": "Node16",
     "noImplicitAny": true,
     "outDir": "dist/",
     "skipLibCheck": true,
```

Linting, `tsc` and tests are all passing now.

File: src/index.ts [hunk 1]
```diff
@@ -2,7 +2,7 @@ import type { AuthForHAR } from './lib/configure-security';
 import type { Extensions } from '@readme/oas-extensions';
 import type { PostDataParams, Request } from 'har-format';
 import type Oas from 'oas';
-import type { SchemaWrapper } from 'oas/dist/operation/get-parameters-as-json-schema';
+import type { SchemaWrapper } from 'oas/operation/get-parameters-as-json-schema';
```
Comment (erunion) [inline src/index.ts:5]: very cool that we no longer have to pull these out of the dist.

**Comment by prisis** [comment]
Hey :), you need to change the "moduleResolution" to "bundler" because you using tsup and not tsc anymore :)

https://devblogs.microsoft.com/typescript/announcing-typescript-5-0/#moduleresolution-bundler

File: src/index.ts (hunk 1 above)
Comment (prisis) [reply src/index.ts:5]: With the new exports key in package.json, you have now a nice public api :)

File: tsup.config.ts [hunk 2]
```diff
@@ -5,15 +5,16 @@ import { defineConfig } from 'tsup';
...
+  entry: ['src/index.ts', 'src/lib/configure-security.ts'],
   format: ['esm', 'cjs'],
-  silent: !options.watch,
   minify: false,
-  dts: true,
+  shims: true,
+  silent: !options.watch,
   sourcemap: true,
-  clean: true,
   splitting: true,
```
Comment (kanadgupta) [inline tsup.config.ts:18]: per [their docs](https://tsup.egoist.dev/#code-splitting), they already code split for ESM by default and this flag is enables an experimental feature where it attempts to split for CJS. i'm pretty hesitant to try that to be honest.

```suggestion
```

File: tsup.config.ts [hunk 3]
```diff
@@ -5,15 +5,16 @@ import { defineConfig } from 'tsup';
...
-  shims: true,
-  cjsInterop: true,
+  treeshake: true,
```
Comment (kanadgupta) [inline tsup.config.ts:19]: per [their docs](https://tsup.egoist.dev/#tree-shaking) treeshaking is enabled by default but this flag defers to rollup for treeshaking. i'm fine with leaving this as is but it feels like bad DX that this flag exists in the first place.

File: tsup.config.ts [hunk 4]
```diff
@@ -5,15 +5,16 @@ import { defineConfig } from 'tsup';
...
-  entry: ['src/index.ts'],
-  treeshake: true,
+
+  cjsInterop: true,
+  clean: true,
+  dts: true,
+  entry: ['src/index.ts', 'src/lib/configure-security.ts'],
   format: ['esm', 'cjs'],
-  silent: !options.watch,
   minify: false,
```
Comment (kanadgupta) [inline tsup.config.ts:14]: don't think we need this, it's the default value ([docs](https://tsup.egoist.dev/#minify-output))

```suggestion
```

File: tsup.config.ts [hunk 5]
```diff
@@ -5,15 +5,16 @@ import { defineConfig } from 'tsup';
 
 export default defineConfig((options: Options) => ({
```
Comment (kanadgupta) [inline tsup.config.ts:6]: i think this type and the import above are redundant:

```suggestion
export default defineConfig((options) => ({
```

**Review by kanadgupta — APPROVED** [review]
LGTM for the most part, just a few comments (mostly nits) on this

File: src/index.ts (hunk 1 above)
Comment (erunion) [reply src/index.ts:5]: unclear, but will look into it after i get this published

File: tsup.config.ts (hunk 3 above)
Comment (erunion) [reply tsup.config.ts:16]: Agreed. I'm going to leave this in place as the rollup-generated dists are a lot cleaner.

#### without / with

Screen Shot 2023-09-08 at 10 42 20 AM

**Pull request merged**